Support QSPI Flash driver for Renesas RA6 devices

thaoluonguw opened the pull request:
Add support for QSPI Flash driver for RA6 devices

- Add QSPI Flash driver support on EK_RA6M5, EK_RA6M4, EK_RA6M3 and EK_RA6E2.

Commit:
drivers: flash: Initial support QSPI Flash driver for Renesas RA6

Add QSPI Flash driver supports for Renesas RA6.

Signed-off-by: Tri Nguyen <tri.nguyen.wj@bp.renesas.com>
Signed-off-by: Thao Luong <thao.luong.uw@renesas.com>
Signed-off-by: Khoa Nguyen <khoa.nguyen.xh@renesas.com>

soburi commented on include/zephyr/drivers/flash/ra_flash_api_extensions.h:
```diff
@@ -13,6 +13,10 @@ enum ra_ex_ops {
+enum qspi_flash_ex_op_types {
+	QSPI_FLASH_EX_OP_RESET = 0,
+	QSPI_FLASH_EX_OP_EXIT_QPI = 1,
+};
```
Please insert blank line.

thaoluonguw replied:
I fixed it. Thank you for your comment.

soburi commented on dts/bindings/mtd/renesas,ra-qspi-nor.yaml:
```diff
@@ -0,0 +1,30 @@
+  reg:
+    required: true
+
+  size:
```
Renesas specific property should have `renesas,` prefix.

de-nordic commented on drivers/flash/flash_qspi_renesas_ra.h:
```diff
@@ -0,0 +1,76 @@
+#define QSPI_CMD_RESET_MEM      (0x99) /* Reset Memory */
+#define Ex_SPI_CMD_READ_ID      (0x9F) /* Extended Spi Read JEDEC ID */
+#define QSPI_CMD_READ_ID        (0xAF) /* Qpi Read JEDEC ID */
+#define QSPI_CMD_READ_SFDP      (0x5A) /* Read SFDP */
```
Most of those defines are actually provided here https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/flash/spi_nor.h, I know the naming is SPI_NOR, but adding additional defines only brings new names for functionally the same things.
You could just use the SPI_NOR defines and define these that are not provided by the header.

thaoluonguw replied:
We changed to use macros in https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/flash/spi_nor.h

de-nordic commented on include/zephyr/drivers/flash/ra_flash_api_extensions.h:
```diff
@@ -13,6 +13,11 @@ enum ra_ex_ops {
 };
 
+enum qspi_flash_ex_op_types {
+	QSPI_FLASH_EX_OP_RESET = 0,
```
There is already system wide id for reset
`FLASH_EX_OP_RESET = 0` https://github.com/zephyrproject-rtos/zephyr/blob/03959a20f7834ffe50a51df2b2e45ef87d1b610b/include/zephyr/drivers/flash.h#L640
and if the meaning is the same, the system wide one should be used.

thaoluonguw replied:
We fixed it as your advice.

de-nordic commented on include/zephyr/drivers/flash/ra_flash_api_extensions.h:
```diff
@@ -13,6 +13,11 @@ enum ra_ex_ops {
 
+enum qspi_flash_ex_op_types {
+	QSPI_FLASH_EX_OP_RESET = 0,
+	QSPI_FLASH_EX_OP_EXIT_QPI = 1,
```
This define is in the same range as system wide definition for non-vendor space, and as such is not allowed.
Please refer to https://github.com/zephyrproject-rtos/zephyr/blob/03959a20f7834ffe50a51df2b2e45ef87d1b610b/include/zephyr/drivers/flash.h#L617-L631
for instruction on how to define vendor specific operation
and to https://github.com/zephyrproject-rtos/zephyr/blob/03959a20f7834ffe50a51df2b2e45ef87d1b610b/include/zephyr/drivers/flash/stm32_flash_api_extensions.h for an example of such definition.

de-nordic replied:
I think that there is also typo and QPI should be QSPI?

thaoluonguw replied:
We updated for "QSPI_FLASH_EX_OP_EXIT_QPI" as your advice.
And "QPI" isn't typo. We mentioned QPI mode (4-4-4).

de-nordic commented on drivers/flash/flash_qspi_renesas_ra.c:
```diff
@@ -0,0 +1,501 @@
+		return 0;
+	}
+
+	if (!qspi_flash_ra_valid(QSPI_NOR_FLASH_SIZE, offset, len)) {
+		return -EINVAL;
+	}
+
+	if (len % QSPI_ERASE_BLK_SZ != 0) {
+		return -EINVAL;
+	}
```
You probably want to have some logging in at least one of the ifs, otherwise you have the same error returned to user with no way to distinguish problems except for debugging.

thaoluonguw replied:
We fixed it. We added log to distinguish for errors.

de-nordic commented on drivers/flash/flash_qspi_renesas_ra.c:
```diff
@@ -0,0 +1,501 @@
+	ARG_UNUSED(out);
+	acquire_device(dev);
+
+	switch (code) {
+	case QSPI_FLASH_EX_OP_EXIT_QPI:
+		if (SPI_FLASH_PROTOCOL_QPI != qspi_data->qspi_cfg.spi_protocol) {
+			err = 0;
+			break;
+		}
+		cmd = QSPI_CMD_EXIT_QPI_MODE;
+		err = R_QSPI_DirectWrite(&qspi_data->qspi_ctrl, &cmd, ONE_BYTE, false);
+		if (err != FSP_SUCCESS) {
+			LOG_ERR("qspi: direct write failed");
+			err = -EIO;
+		}
+		break;
+
+	case QSPI_FLASH_EX_OP_RESET:
+		cmd = QSPI_CMD_RESET_EN;
+		err = R_QSPI_DirectWrite(&qspi_data->qspi_ctrl, &cmd, ONE_BYTE, false);
+		if (err == FSP_SUCCESS) {
+			cmd = QSPI_CMD_RESET_MEM;
+			err = R_QSPI_DirectWrite(&qspi_data->qspi_ctrl, &cmd, ONE_BYTE, false);
+			if (err != FSP_SUCCESS) {
+				LOG_ERR("qspi: direct write failed");
+				err = -EIO;
+			}
```
You lack here userspace processing, is that desired?

thaoluonguw replied:
Yes, it is designed without userspace processing.
It is supported to reset QPI mode and memory.

de-nordic commented on drivers/flash/flash_qspi_renesas_ra.c:
```diff
@@ -0,0 +1,501 @@
+		return -EINVAL;
+	}
+
+	if (!len) {
```
Here you fail operation of -EINVAL, even if len == 0 would cause nothing happening at all,
on the other side in `qspi_flash_ra_write` you first check len == 0 and bypass any more checks as there is no more work to.

Consider having one flow everywhere, for example checking the  len == 0 and just exit as there is no work to do and actually no possibility to failure

de-nordic reviewed (changes requested):
Issue with EX_OP and mostly nits regarding other areas.

de-nordic commented:
There is a lot of QPI in the code, shouldn't that be QSPI?

thaoluonguw commented on drivers/flash/flash_qspi_renesas_ra.h:
```diff
@@ -0,0 +1,76 @@
+
+#define QSPI_CMD_RESET_EN       (0x66) /* Reset Enable */
+#define QSPI_CMD_RESET_MEM      (0x99) /* Reset Memory */
+#define Ex_SPI_CMD_READ_ID      (0x9F) /* Extended Spi Read JEDEC ID */
```
We fixed it. We changed to use macros in https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/flash/spi_nor.h

thaoluonguw commented:
> There is a lot of QPI in the code, shouldn't that be QSPI?

@de-nordic : Thank you so much for your review. We fixed issues which you pointed out.
Please check the updates at: https://github.com/zephyrproject-rtos/zephyr/compare/81878488837d751b4711ff629a61c9c098011dce..f462ce16a76f505fde14c40293db08f204f108a2 

For "QPI" in the code, it isn't typo of "QSPI", this is support for QPI mode (4-4-4).

Could you please help us to recheck for the update? Thanks again.

quytranpzz committed:
dts: renesas: Add devicetree property for QSPI support on RA6

Add qspi node on Renesas RA6 devicetree to support QSPI flash driver

Signed-off-by: Quy Tran <quy.tran.pz@renesas.com>

quytranpzz committed:
boards: renesas: Add boards support for QSPI flash driver

Add support for QSPI flash driver on EK-RA6E2, EK-RA6M3, EK-RA6M4
and EK-RA6M5

Signed-off-by: Quy Tran <quy.tran.pz@renesas.com>
Signed-off-by: Khoa Nguyen <khoa.nguyen.xh@renesas.com>

quytranpzz commented:
rebases to resolve conflicts and separate the commits for clarity

khoa-nguyen-18 commented:
rebase main to solve conflicts

soburi commented on drivers/flash/Kconfig:
```diff
@@ -205,6 +205,8 @@ source "drivers/flash/Kconfig.stm32_xspi"
 source "drivers/flash/Kconfig.xmc4xxx"
 # zephyr-keep-sorted-stop
 
+source "drivers/flash/Kconfig.renesas_ra_flash_qspi"
```
Why put it outside of `# zephyr-keep-sorted-start` and `# zephyr-keep-sorted-stop`?

thaoluonguw replied:
It's mistake. I updated it. Thank you so much for your comment.

soburi commented on drivers/flash/CMakeLists.txt:
```diff
@@ -154,6 +154,7 @@ zephyr_library_include_directories_ifdef(
   ${ZEPHYR_BASE}/drivers/memc
 )
 
+zephyr_library_sources_ifdef(CONFIG_FLASH_QSPI_RENESAS_RA flash_qspi_renesas_ra.c)
```
Shouldn't it be after https://github.com/zephyrproject-rtos/zephyr/pull/78959/files#diff-2622b0ce0ec8ab10c395ae4c3ca5e6aa64d3e8943d6dd340ca7d2f6921fb9c2cR38 ?

soburi commented on drivers/flash/flash_qspi_renesas_ra.c:
```diff
@@ -0,0 +1,566 @@
+	struct st_spi_flash_cfg qspi_cfg;
+	struct k_sem sem;
+};
+struct ra_qspi_nor_flash_config {
```
insert blank line

soburi commented on drivers/flash/flash_qspi_renesas_ra.c:
```diff
@@ -0,0 +1,566 @@
+	{.command = 0xC7, .size = SPI_FLASH_ERASE_SIZE_CHIP_ERASE},
+};
+
+#define ERASE_COMMAND_LENGTH(arr) (sizeof(arr) / sizeof((arr)[0]))
```
Let's put all the macro definitions at the top.

soburi commented on drivers/flash/flash_qspi_renesas_ra.c:
```diff
@@ -0,0 +1,566 @@
+PINCTRL_DT_DEFINE(QSPI0_NODE);
+LOG_MODULE_REGISTER(flash_qspi_renesas_ra, CONFIG_FLASH_LOG_LEVEL);
+
+#define QSPI_ENABLE_QUAD_MODE DT_PROP(RA_QSPI_NOR_NODE, qpi_enable)
```
ditto

thaoluonguw commented:
rebase main to solve conflicts

thaoluonguw commented:
@de-nordic, @soburi : Could you please recheck again ? Thank you so much.

thaoluonguw commented:
Update to get size from reg.

thaoluonguw commented:
Rebase to solve conflict.

khoa-nguyen-18 commented:
Just rebased the main branch to resolve conflicts. The PR will need some updates, which will be pushed soon.

khoa-nguyen-18 committed:
tests: drivers: flash: Add support to test qspi_nor for RA6

Add support to test flash/common for qspi_nor on Renesas
ek_ra6m5, ek_ra6m4, ek_ra6m3, ek_ra6e2

Signed-off-by: Khoa Nguyen <khoa.nguyen.xh@renesas.com>

khoa-nguyen-18 committed:
samples: drivers: spi_flash: add support for qspi Renesas RA

Add support for qspi Renesas RA to run the sample spi_flash

Signed-off-by: Khoa Nguyen <khoa.nguyen.xh@renesas.com>

khoa-nguyen-18 committed:
samples: drivers: jesd216: Add support for qspi_nor for RA

Add support for qspi_nor for the sample jesd216 to run sample
on Renesas RA6 boards

Signed-off-by: Khoa Nguyen <khoa.nguyen.xh@renesas.com>

khoa-nguyen-18 commented:
- Update source driver file name
- Add commit to remove conf file to not test `get_size` case for flash and qspi on test app `flash/common` for all Renesas boards
- Add test case in `tests/drivers/flash/common/testcase.yaml` to test qspi

khoa-nguyen-18 commented:
Hi @de-nordic, @soburi, @duynguyenxa, @rghaddab can we restart the review of this PR?

khoa-nguyen-18 commented:
just rebase to solve the conflict

khoa-nguyen-18 commented:
modify the naming of `ra-qspi-nor` -> `qspi-nor-flash`

khoa-nguyen-18 commented:
rebase to solve conflicts

khoa-nguyen-18 commented:
rebase main to solve conflicts

khoa-nguyen-18 commented:
rebase main to solve conflicts

facchinm commented:
Hi @khoa-nguyen-18 , could you rebase again upon current `main` ? I'd love to complete https://github.com/zephyrproject-rtos/zephyr/pull/85337 and this is one of the few missing parts :slightly_smiling_face:

khoa-nguyen-18 commented:
- Rebase to resolve conflicts
- Update support for flash/common related to Renesas QSPI

khoa-nguyen-18 commented:
Hi @de-nordic, could you please help confirm your requested changes in this PR?

pillo79 commented:
@khoa-nguyen-18 friendly notice - this PR needs a rebase to fix a merge conflict.

khoa-nguyen-18 commented:
> @khoa-nguyen-18 friendly notice - this PR needs a rebase to fix a merge conflict.

Thanks for reminding me. I’ve rebased on main to fix the conflict.

khoa-nguyen-18 commented:
Hi @de-nordic, @KhiemNguyenT, @soburi, @duynguyenxa, @thenguyenyf, could you please take a look at this PR and help determine whether it can be merged into 4.2?

khoa-nguyen-18 commented:
Hi @de-nordic, just a gentle reminder, we're still waiting for your feedback. Please take a look at this PR when you get a chance.

khoa-nguyen-18 commented:
Hi @de-nordic, we have enough review approvals, but we’re still missing yours. Could you please take a look at this PR?

kartben commented:
> Hi @kartben, sorry to bother you. Since de-nordic might be quite busy at the moment, could you help us move this PR forward?

I think he's on vacation. I will try to have a look

kartben commented on include/zephyr/drivers/flash/ra_flash_api_extensions.h:
```diff
@@ -11,6 +11,9 @@
 enum ra_ex_ops {
 	FLASH_RA_EX_OP_WRITE_PROTECT = FLASH_EX_OP_VENDOR_BASE,
+
+	/* Reset Flash device (at QPI(4-4-4) mode) */
+	QSPI_FLASH_EX_OP_EXIT_QPI,
```
this needs proper documentation, see FLASH_RA_EX_OP_WRITE_PROTECT doc in `main` even if I guess there's no `in` / `out` to document, there should still be proper doxygen documentation

kartben commented on drivers/flash/flash_renesas_ra_qspi.c:
```diff
@@ -0,0 +1,575 @@
+#define PAGE_SIZE_BYTE       SPI_NOR_PAGE_SIZE
+/* sector size of QSPI flash device */
+#define BLOCK_SIZE_4K        (4096U)
+#define BLOCK_SIZE_32K       (32678U)
```
+1

khoa-nguyen-18 replied:
Nice catch, I have fixed this

kartben commented on drivers/flash/flash_renesas_ra_qspi.c:
```diff
@@ -0,0 +1,575 @@
+{
+	struct qspi_flash_ra_data *qspi_data = dev->data;
+	int err = 0;
+	uint8_t buffer[size];
+
+	acquire_device(dev);
+	memset(&buffer[0], 0, sizeof(buffer));
+	buffer[0] = QSPI_QPI_CMD_RDSFDP;
+	buffer[1] = addr;
+	buffer[2] = addr >> 8;
+	buffer[3] = addr >> 16;
+
+	err = R_QSPI_DirectWrite(&qspi_data->qspi_ctrl, &buffer[0], FOUR_BYTE, true);
+	if (err != FSP_SUCCESS) {
+		LOG_ERR("Direct write for READ SFDP failed");
+		err = -EIO;
+		goto out;
+	}
+
+	err = R_QSPI_DirectRead(&qspi_data->qspi_ctrl, &buffer[0], size);
```
https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#:~:text=Variable%2Dlength%20array%20types%20shall%20not%20be%20used

khoa-nguyen-18 replied:
My mistake, I have updated like the suggestion

kartben commented on drivers/flash/flash_renesas_ra_qspi.c:
```diff
@@ -0,0 +1,575 @@
+#define BLOCK_SIZE_32K       (32678U)
+#define BLOCK_SIZE_64K       (65536U)
+/* QSPI flash address through page*/
+#define QSPI_FLASH_ADDRESS(page_no)                                                                \
+	((uint8_t *)(QSPI_DEVICE_START_ADDRESS + ((page_no) * SPI_NOR_PAGE_SIZE)))
```
unused?

khoa-nguyen-18 replied:
Yes, I have removed it. Many thanks

kartben commented:
@khoa-nguyen-18 @thaoluonguw ping in case you missed my comments
